### PR TITLE
 Skip empty row groups in parquet reader

### PR DESCRIPTION
**Pull request by qqibrow**
This PR fixes issue #7778 

The commit makes velox parquet reader able to skip empty row groups, instead of crash. This behavior is inline with other parquet readers, e.g, presto parquet reader and parquet-format. 

Additionally, this PR simplifies the code by removing the variable `skippedRowGroups_`

### CODE REVIEW

**Comment by qqibrow** [comment]
@yingsu00 could you take a look? thanks!

**Review by yingsu00 — APPROVED** [review]

**Comment by yingsu00** [comment]
@qqibrow Can you please check the failed tests?

**Comment by qqibrow** [comment]
thanks for the review! Let me rebase and re-run the test.

**Comment by qqibrow** [comment]
I don't think this error message is relevant:
```
FAILED: velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ArithmeticFunctionsRegistration.cpp.o 
/usr/bin/ccache /opt/rh/gcc-toolset-9/root/bin/g++  -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CHRONO_DYN_LINK -DBOOST_CHRONO_NO_LIB -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_EXPORT=EXPORT -DBOOST_CONTEXT_NO_LIB -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK=1 -DBOOST_FILESYSTEM_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DBOOST_THREAD_USE_DLL -DFOLLY_HAVE_INT128_T=1 -DGFLAGS_IS_A_DLL=0 -DGLOG_CUSTOM_PREFIX_SUPPORT -DSIMDJSON_THREADS_ENABLED=1 -I_deps/protobuf-src/src -I../../. -I../../velox/external/xxhash -I_deps/folly-src -I_deps/folly-build -I_deps/gflags-build/include -I_deps/glog-build -I/usr/include/libdwarf -I_deps/fmt-src/include -I_deps/boost-src/libs/crc/include -I_deps/boost-src/libs/array/include -I_deps/boost-src/libs/assert/include -I_deps/boost-src/libs/config/include -I_deps/boost-src/libs/core/include -I_deps/boost-src/libs/static_assert/include -I_deps/boost-src/libs/throw_exception/include -I_deps/boost-src/libs/integer/include -I_deps/boost-src/libs/type_traits/include -I_deps/boost-src/libs/circular_buffer/include -I_deps/boost-src/libs/concept_check/include -I_deps/boost-src/libs/preprocessor/include -I_deps/boost-src/libs/move/include -I_deps/boost-src/libs/math/include -I_deps/boost-src/libs/lexical_cast/include -I_deps/boost-src/libs/container/include -I_deps/boost-src/libs/intrusive/include -I_deps/boost-src/libs/container_hash/include -I_deps/boost-src/libs/describe/include -I_deps/boost-src/libs/mp11/include -I_deps/boost-src/libs/winapi/include -I_deps/boost-src/libs/predef/include -I_deps/boost-src/libs/numeric/conversion/include -I_deps/boost-src/libs/conversion/include -I_deps/boost-src/libs/smart_ptr/include -I_deps/boost-src/libs/typeof/include -I_deps/boost-src/libs/mpl/include -I_deps/boost-src/libs/utility/include -I_deps/boost-src/libs/io/include -I_deps/boost-src/libs/range/include -I_deps/boost-src/libs/detail/include -I_deps/boost-src/libs/iterator/include -I_deps/boost-src/libs/function_types/include -I_deps/boost-src/libs/fusion/include -I_deps/boost-src/libs/tuple/include -I_deps/boost-src/libs/functional/include -I_deps/boost-src/libs/function/include -I_deps/boost-src/libs/bind/include -I_deps/boost-src/libs/type_index/include -I_deps/boost-src/libs/optional/include -I_deps/boost-src/libs/regex/include -I_deps/boost-src/libs/random/include -I_deps/boost-src/libs/dynamic_bitset/include -I_deps/boost-src/libs/system/include -I_deps/boost-src/libs/variant2/include -I_deps/boost-src/libs/multi_index/include -I_deps/boost-src/libs/foreach/include -I_deps/boost-src/libs/serialization/include -I_deps/boost-src/libs/spirit/include -I_deps/boost-src/libs/endian/include -I_deps/boost-src/libs/phoenix/include -I_deps/boost-src/libs/proto/include -I_deps/boost-src/libs/pool/include -I_deps/boost-src/libs/thread/include -I_deps/boost-src/libs/atomic/include -I_deps/boost-src/libs/chrono/include -I_deps/boost-src/libs/ratio/include -I_deps/boost-src/libs/rational/include -I_deps/boost-src/libs/date_time/include -I_deps/boost-src/libs/algorithm/include -I_deps/boost-src/libs/exception/include -I_deps/boost-src/libs/unordered/include -I_deps/boost-src/libs/tokenizer/include -I_deps/boost-src/libs/variant/include -I_deps/boost-src/libs/multiprecision/include -I_deps/boost-src/libs/uuid/include -I_deps/boost-src/libs/tti/include -I_deps/boost-src/libs/context/include -I_deps/boost-src/libs/filesystem/include -I_deps/boost-src/libs/program_options/include -I_deps/boost-src/libs/any/include -I_deps/glog-src/src -I_deps/xsimd-src/include -I_deps/re2-src -I_deps/simdjson-src/include -isystem ../../velox -isystem ../../velox/external -mavx2 -mfma -mavx -mf16c -mlzcnt -std=c++17 -mbmi2 -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-implicit-fallthrough          -Wno-empty-body          -Wno-class-memaccess          -Wno-comment          -Wno-int-in-bool-context          -Wno-redundant-move          -Wno-array-bounds          -Wno-maybe-uninitialized          -Wno-unused-result          -Wno-format-overflow          -Wno-strict-aliasing          -Wno-type-limits          -Wno-stringop-overflow          -Wno-stringop-overread          -Wno-return-type -Werror -g -fPIC   -fdiagnostics-color=always -Wno-deprecated-declarations -pthread -std=gnu++17 -MD -MT velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ArithmeticFunctionsRegistration.cpp.o -MF velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ArithmeticFunctionsRegistration.cpp.o.d -o velox/functions/prestosql/registration/CMakeFiles/velox_functions_prestosql.dir/ArithmeticFunctionsRegistration.cpp.o -c ../../velox/functions/prestosql/registration/ArithmeticFunctionsRegistration.cpp
g++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
``` 
https://app.circleci.com/pipelines/github/facebookincubator/velox/41853/workflows/a6fb49b2-fd16-4c3d-8008-a19684e7c501/jobs/288533
I might have seen similar error when I use large NUM_THREAD to build. @yingsu00 WDYT?

**Comment by yingsu00** [comment]
@qqibrow Can you try changing the 16 to 8 here?

https://github.com/facebookincubator/velox/blob/7376fb29f72fa3109e9b0eb370adc9ef73a20884/.circleci/dist_compile.yml#L256C37-L256C37

**Comment by yingsu00** [comment]
@Yuhta Could you please review and merge this? Thanks!

**Comment by yingsu00** [comment]
@qqibrow Can you just squash the commits into 1?

**Comment by qqibrow** [comment]
@yingsu00 done. thanks!

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 1]
```diff
@@ -698,14 +698,15 @@ void ParquetRowReader::filterRowGroups() {
     auto rowGroupInRange =
         (fileOffset >= options_.getOffset() &&
          fileOffset < options_.getLimit());
+
+    auto inBlacklist =
```
[the remark this answers is not in the document]
Comment (chliang71) [reply velox/dwio/parquet/reader/ParquetReader.cpp:702]: or rename to "isExcluded" following the comment below?

[the remark this answers is not in the document]
Comment (qqibrow) [reply velox/dwio/parquet/reader/ParquetReader.cpp:702]: sounds good

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 2]
```diff
@@ -698,14 +698,15 @@ void ParquetRowReader::filterRowGroups() {
     auto rowGroupInRange =
         (fileOffset >= options_.getOffset() &&
          fileOffset < options_.getLimit());
+
+    auto isExcluded =
+        (i < res.totalCount && bits::isBitSet(res.filterResult.data(), i));
+
+    auto isEmpty = rowGroups_[i].num_rows == 0;
     // A skipped row group is one that is in range and is in the excluded list.
```
Comment (pedroerp) [inline velox/dwio/parquet/reader/ParquetReader.cpp:706]: missed this in the first round, but can you update the comment too?

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 3]
```diff
@@ -698,14 +698,15 @@ void ParquetRowReader::filterRowGroups() {
     auto rowGroupInRange =
         (fileOffset >= options_.getOffset() &&
          fileOffset < options_.getLimit());
+
+    auto isExcluded =
```
Comment (pedroerp) [inline velox/dwio/parquet/reader/ParquetReader.cpp:702]: nit: for readability, merge this two in a single code block and add blank line afterwards

```
auto isExcluded = (i < res.totalCount && bits::isBitSet(res.filterResult.data(), i));
auto isEmpty = rowGroups_[i].num_rows == 0;

if (...) {
```

**Review by pedroerp — COMMENTED** [review]
sorry, two last small nits then I'll get it merged :)

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 3 above)
Comment (qqibrow) [reply velox/dwio/parquet/reader/ParquetReader.cpp:702]: The first statement rollovers to next line by `format-fix`

**Pull request closed without merging**